### ISSUE TITLE
Please allow mapping of staged flippers in VPX

### ISSUE THREAD

**Issue by Kamun** [vpinball/vpinball#651]
**Is your feature request related to a problem? Please describe.**

I had this all written but my phone backed out of the page and I lost it all. 

I am unable to use the staged flipper mods I installed on my cab because the buttons are not assignable directly on vpx. I must use j2k to assign my button to the keyboard key to which the upper flippers are assigned. 

**Describe the solution you'd like**

Please allow direct button assignment via the keys/buttons/nudge options menu in VPX for the upper flippers, left and right. 

**Describe alternatives you've considered**

I have considered utilizing j2k however I do not wish to install a separate program that can contribute to input lag just for two buttons that should be assignable to work. 

**Additional context**

**Comment by gear323** [comment]
you can use the default keys for the upper flippers or you can edit which keys are used by manually editing VPMkeys.vbs

Const keyUpperLeft	   = 30  '(A)		Upper Left Flipper button
Const keyUpperRight	   = 40  '(')		Upper Right Flipper button 


Its a workaround for now until they are supported better in the GUI

**Comment by Kamun** [comment]
This does not allow you to map these flippers to your buttons directly. There is no way to tell vpx that "Button 12" (my right staged flipper button) is the upper right flipper, using that script file, unless I'm not understanding it correctly.

**Comment by wiesshund** [comment]
Machines with staged flips generally don't have magnasave buttons.
So the staged flipping is usually run by the  magnasave as far as romless tables go

for ROM tables, in vpm keys
' Double leaf/staged flipper switches: By default these are just set to LeftFlipperKey and RightFlipperKey
Dim keyStagedFlipperL : keyStagedFlipperL = LeftMagnasave
Dim keyStagedFlipperR : keyStagedFlipperR = RightMagnasave

would accomplish that.

**Comment by wiesshund** [comment]
then of course, set your staged to the magnasaves

**Comment by wiesshund** [comment]
If you do what gear is suggesting
1st you have to have your controller send keyboard codes for those buttons, if the controller does keyboard emulation.
If it does not, you can use antimicro to do it (better than joy2key)

and then i think 
' Double leaf/staged flipper switches: By default these are just set to LeftFlipperKey and RightFlipperKey
Dim keyStagedFlipperL : keyStagedFlipperL = keyUpperLeft
Dim keyStagedFlipperR : keyStagedFlipperR = keyUpperRight

**Comment by Kamun** [comment]
I have magnas and staged flippers, but I do not wish to use j2k or other programs to emulate keyboard control, which my controller does not use. Ideally I would like to be able to assign my staged flips, which are buttons 11/12 on my controller, just like I can assign left flipper to button 5 for example, and it just works.

If this were possible with just changing the number in vpmkeys this would work but in my experience that script isn't looking at joystick button assignments, just keyboard assignment. Why is there no native staged flipper button assignment support for a cab setup without having to emulate a keyboard? :(

**Comment by wiesshund** [comment]
i guess it wasnt included because the amount of machines employing a staged flipper button
and the number of VP users outfitting the cabs in that fashion was small
And 
The number of machines having Mangasave is really small, and they dont use staged flips
So most people just used the staged flip as the magnasave.

I mean you could use the custom key, there are 4 of them in vpx
i forget it's vpmkeys name off hand, toxie would know

There is a way to do it though as the few other people with same setup as yours have been doing it

**Comment by Kamun** [comment]
Can this please be included as an official binding? Rebinding the flippers to the magna is not a solution and I do not wish to use a separate keyboard binding program to set this up. Staged flippers are a very common feature in many real life tables and to not be able to easily replicate this feature without third party software is a huge shortcoming imo.

**Comment by Lovinity** [comment]
I too can attest that having a specific binding for staged flippers without having to use complicated third party software is a popular demand in the VP community. I can also attest that using other key bindings not meant for staged flippers is not a solution. We cannot, and never should, assume that a table won't also use the other keys such as magna save, because it can / will happen if there isn't already a table out there.

Staged flippers are very common in tables. And lots of v players want the ability to use, and practice with, staged flippers in VPX.

**Comment by wiesshund** [comment]
did you try binding to the custom keys in vpx?
i know some people were using that.
It isnt perfect, but it's something until a nicer solution can be added

**Comment by Kamun** [comment]
Binding to custom keys was one of the first things I tried. Can't seem to get the table script to respond to the key being pressed though.

**Comment by wiesshund** [comment]
you have to assign the custom keys to the staged flips in vpmkeys
unless you are talking an original romless table?

Inside VPX script i think they are known as CustomKey1 etc
@toxieainc can advise as to what name vpmkeys would refer to them as?
i dont recall offhand

**Comment by Kamun** [comment]
When assigned in this fashion the staged flippers would not respond to being pressed. The custom keys assignments do not work as expected when trying this previously. It would work much better if this were directly assignable via vpx and allow buttons to be mapped directly via the button assignment menu.

**Comment by wiesshund** [comment]
They do work
I just do not remember what to tell you to refer to them as in vpmkeys
@toxieainc would know the right answer
Unless i can get someone that had set it up that way to respond back to me

it is not a perfect solution
i use those keybindings for the 4 joystick directions in things like baby pacman, which they are kind of more suited for.
staged flip left and staged flip right dedicated bindings, and then making sure vpn understands them
would be the proper way, like you are saying.

I will shout out again and see if i can get someone to answer back on what stagedflips looks like in vpmkeys
along with their vpx keybindings

**Comment by Primetime5k** [comment]
> Machines with staged flips generally don't have magnasave buttons. So the staged flipping is usually run by the magnasave as far as romless tables go

Grand Lizard and Solar Fire have both staged flippers and magnas.  And romless tables don't dictate if magnas are used for staging.  Blood Machines and Iron Maiden are both Romless and use the 2nd stage on the dual-leaf switches.   I know Shovel Knight did the magna/staged thing, but that can easily be tweaked to use the 2nd stage vs. magnas.

**Comment by Primetime5k** [comment]
I've tried binding the custom keys and was never able to get it to work, but probably overlooking something.  Agree with Kamun - seems like it isn't a major lift to be able to give the capability via the GUI, but I'd be happy to figure out how to manually tweak it in order to remove J2K from the mix.

**Comment by wiesshund** [comment]
> > Machines with staged flips generally don't have magnasave buttons. So the staged flipping is usually run by the magnasave as far as romless tables go
> 
> Grand Lizard and Solar Fire have both staged flippers and magnas. And romless tables don't dictate if magnas are used for staging. Blood Machines and Iron Maiden are both Romless and use the 2nd stage on the dual-leaf switches. I know Shovel Knight did the magna/staged thing, but that can easily be tweaked to use the 2nd stage vs. magnas.

Which is not the general population.

And i dont know what you mean about romless
I can script a table to know the difference between staged keys and magnasave and flippers and use all 3
I do not know that anyone has as of yet though? but you most definitely can dictate it in a romless table

**Comment by wiesshund** [comment]
> I've tried binding the custom keys and was never able to get it to work, but probably overlooking something. Agree with Kamun - seems like it isn't a major lift to be able to give the capability via the GUI, but I'd be happy to figure out how to manually tweak it in order to remove J2K from the mix.

Yea sorry :(
No one is getting back to me on what vpmkeys refers to the custom keys as
and you have to have the right name in order to direct the staged flips to look at them.

let me look at something for you.
what joy buttons do your staged switches equate to?
This may take a bit

**Comment by Kamun** [comment]
> > I've tried binding the custom keys and was never able to get it to work, but probably overlooking something. Agree with Kamun - seems like it isn't a major lift to be able to give the capability via the GUI, but I'd be happy to figure out how to manually tweak it in order to remove J2K from the mix.
> 
> Yea sorry :( No one is getting back to me on what vpmkeys refers to the custom keys as and you have to have the right name in order to direct the staged flips to look at them.
> 
> let me look at something for you. what joy buttons do your staged switches equate to? This may take a bit

Buttons 11 and 12

**Comment by wiesshund** [comment]
If i knew what i was doing, i would try looking at the source and making a test branch

10 years ago this used to work

`'Last Updated in VBS v3.56

'Key codes that can be used in the mappings below:
'0			11		L CONTROL		29
'1			2		LEFT ARROW		203
'2			3		L SHIFT			42
'3			4		L WIN			219
'4			5		M			50
'5			6		MINUS			12
'6			7		MULTIPLY		55
'7			8		N			49
'8			9		NUM LOCK		69
'9			10		NUMPAD 0		82
'A			30		NUMPAD 1		79
'ADD			78		NUMPAD 2		80
'APOSTROPHE 		40		NUMPAD 3		81
'AT			145		NUMPAD 4		75
'AX			150		NUMPAD 5		76
'B			48		NUMPAD 6		77
'BACK			14		NUMPAD 7		71
'BACKSLASH		43		NUMPAD 8		72
'BACKSPACE		14		NUMPAD 9		73
'C			46		NUMPAD COMMA		179
'CALCULATOR 		161		NUMPAD ENTER		156
'CAPS LOCK		58		NUMPAD EQUALS		141
'COLON			146		NUMPAD MINUS		74
'COMMA			51		NUMPAD PERIOD		83
'D			32		NUMPAD PLUS		78
'DECIMAL		83		NUMPAD SLASH		181
'DELETE			211		NUMPAD ASTERISK 	55
'DIVIDE			181		O			24
'DOWN			208		P			25
'DOWN ARROW 		208		PAUSE			197
'E			18		PERIOD			52
'END			207		PAGE DOWN		209
'EQUALS			13		PAGE UP			201
'ESCAPE			1		Q			16
'F			33		R			19
'F1			59		R ALT			184
'F2			60		R BRACKET		27
'F3			61		R CTRL			157
'F4			62		RETURN			28
'F5			63		RIGHT			205
'F6			64		RIGHT ARROW		205
'F7			65		R MENU			184
'F8			66		R SHIFT			54
'F9			67		R WIN			220
'F10			68		S			31
'F11			87		SCROLL			70
'F12			88		SEMICOLO		39
'F13			100		SLASH			53
'F14			101		SPACE			57
'F15			102		STOP			149
'G			34		T			20
'H			35		TAB			15
'HOME			199		U			22
'I			23		UNDERLINE		147
'INSERT			210		UP ARROW		200
'J			36		V			47
'K			37		W			17
'L			38		X			45
'L ALT			56		Y			21
'L BRACKET		26		Z			44

Option Explicit

'
' This is a list of standard keys used by *.vbs files
'

' This is an addition of joystick button mappings
Const joy1                = 16777264  '----
Const joyUp             = 16777264  '       ---
Const joy2                = 16777265  '             --
Const joyDown         = 16777266  '                   .  Buttons 1-4 &
Const joy3               = 16777266  '                    .  Stick share values 
Const joyLeft           = 16777267  '              -- 
Const joy4               = 16777267  '       ---
Const joyRight        = 16777265   '----
Const joy5              = 16777268  ' button 5
Const joy6              = 16777269  ' button 6
Const joy7              = 16777270  ' button 7
Const joy8              = 16777271  ' button 8
Const joy9              = 16777272  ' button 9
Const joy10            = 16777273  ' button 10
Const joy11            = 16777274  ' button 11
Const joy12            = 16777275  ' button 12


' Double leaf/staged flipper switches: By default these are just set to LeftFlipperKey and RightFlipperKey
Dim keyStagedFlipperL : keyStagedFlipperL = LeftFlipperKey
Dim keyStagedFlipperR : keyStagedFlipperR = RightFlipperKey

Const keyShowOpts	= 59  '(F1)		Show options
Const keyShowKeys	= 60  '(F2)		Show Keys
Const keyReset		= 61  '(F3)		Reset Emulation
Const keyFrame		= 62  '(F4)		Toggle Window Lock
Const keyDoubleSize	= 63  '(F5)		Toggle displaysize
Const keyShowDips	= 64  '(F6)		Show Dip Switch / Options Menu
Const keyAddBall	= 76  '(NUMPAD 5)	Add extra ball
Const keyBangBack	= 20  '(T)		Bang Back

Const keyInsertCoin1	 = 79   '(NUMPAD 1)		Insert coin in slot 1
Const keyInsertCoin2	 = 80   '(NUMPAD 2)		Insert coin in slot 2
Const keyInsertCoin3	 = 81   '(NUMPAD 3)		Insert coin in slot 3
Const keyInsertCoin4	 = 77   '(NUMPAD 6)		Insert coin in slot 4
Const keySlamDoorHit	 = 199 '(Home)	Slam Tilt
Const keyCoinDoor		 = 207 '(End)	Opening/Closing Coin Door
Const toggleKeyCoinDoor	 = True '		If true then a key press/switch change will flip the state from on/off, if false then one will have the real coindoor behaviour (for cabinet setups with real coindoors)
Const inverseKeyCoinDoor = False '		If false then the coindoor switch behaves like a original coindoor (key pressed = closed, key not pressed = open), otherwise the switch is inverted (key pressed = open, key not pressed = closed)

Const keyCancel		 	= 8   '(7)		Coin Door Cancel Button (WPC)
Const keyDown			= 9   '(8)		Coin Door Down Button (WPC)
Const keyUp			= 10  '(9)		Coin Door Up Button (WPC)
Const keyEnter		 	= 11  '(0)		Coin Door Enter Button (WPC)
Const keyAdvance		= 9   '(8)		Coin Door Advance Button
Const keyUpDown		 	= 8   '(7)		Coin Door Up/Down Button
Const keyCPUDiag		= 10  '(9)		CPU Diagnostic Button
Const keySoundDiag	 	= 11  '(0)		Sound CPU Diagnostic Button
Const keyMasterEnter		= 201 '(PgUp)		Master Enter
Const keyHiScoreReset		= 7   '(6)		Reset Hiscores
Const keySelfTest		= 8   '(7)
Const keyBlack		 	= 8   '(7)		Sega/Data East/Stern
Const keyGreen		 	= 9   '(8)		Sega/Data East/Stern
Const keyRed			= 10  '(9)		Sega/Data East/Stern
Const keyKPGame		 	= 7   '(6)		Bally 6803
Const keyKPEnter		= 156 '(NumpadEnter)
Const keyVPMVolume	 	= 88  '(F12)	Added Control to set VPM Volume
Const keyConfigurations = 12 '(-)		Taito

' Optional keys
' These are not used nor handled by the vbs files
' They are only here to provide consistent keys for the games.

Const keyUpperLeft	   = 30  '(A)			Upper Left Flipper button
Const keyUpperRight	   = 40  '(')			Upper Right Flipper button
Const keyFront		   = 75   '(NUMPAD 4)		Front button (e.g. buy-in)
Const keyRules		   = 19  '(R)			Show rules
Const keyJoyUp		   = 200 '(Up arrow)
Const keyJoyDown	   = 208 '(Down arrow)
Const keyJoyLeft	   = 203 '(Left arrow)
Const keyJoyRight	   = 205 '(Right arrow)
Const keyFire1		   = 56  '(Left CTRL)
Const keyFire2		   = 184 '(Right Alt)
Const keyGameSpecific1 = 16  '(Q)
Const keyGameSpecific2 = 17  '(W)
Const keyGameSpecific3 = 18  '(E)
Const keyGameSpecific4 = 19  '(R)
`

it will barf now, if you try to assign anything to the declared joy# inputs
it will say unable to read vpmkeys.vpn, which i assume means vpx read it, and had a stroke over the command
you can try it, its harmless, the error will stop when you put vpmkeys back to referring to a key code again

I do not know why though
See they used to return values of 16777271  etc when pressed

I am still looking for a solution for you though

**Comment by wiesshund** [comment]
BOOM!
Lightning has struck!

Step one
go to %appdat%\VPinballX
open VpinballX.ini

jump down to customjoykey

edit to read this
joycustom1key = 22
joycustom2key = 23
joycustom3key = 24
joycustom4key = 25

This is U I O and P
if those keys are no good for you, then just pick 4 keys you do not use for vpx or front end
look up the key codes in vpmkeys.vbs

AND LEAVE Vpmkeys.vbs open cause we are going there in a minute

Ok now open VPX, no table just VPX
map custom 1 2 3and 4 as you see fit
Lets say the order is staged left, staged right, something else if you want, one last thing if you want.
if you dont have 2 other things, leave 3 and 4 alone

Ok now in VPM keys
Dim keyStagedFlipperL : keyStagedFlipperL = 22
Dim keyStagedFlipperR : keyStagedFlipperR = 23

If you used keys other than U and I then of course change to the appropriate value.

BAM!
Staged flippers

**Comment by Primetime5k** [comment]
> BOOM! Lightning has struck!
> 
> Step one go to %appdat%\VPinballX open VpinballX.ini
> 
> jump down to customjoykey
> 
> edit to read this joycustom1key = 22 joycustom2key = 23 joycustom3key = 24 joycustom4key = 25
> 
> This is U I O and P if those keys are no good for you, then just pick 4 keys you do not use for vpx or front end look up the key codes in vpmkeys.vbs
> 
> AND LEAVE Vpmkeys.vbs open cause we are going there in a minute
> 
> Ok now open VPX, no table just VPX map custom 1 2 3and 4 as you see fit Lets say the order is staged left, staged right, something else if you want, one last thing if you want. if you dont have 2 other things, leave 3 and 4 alone
> 
> Ok now in VPM keys Dim keyStagedFlipperL : keyStagedFlipperL = 22 Dim keyStagedFlipperR : keyStagedFlipperR = 23
> 
> If you used keys other than U and I then of course change to the appropriate value.
> 
> BAM! Staged flippers

Lol! I was trying to find that ini in the vpin folder - no wonder I didn't find it.  Nice job and TY!

**Comment by wiesshund** [comment]
Test it tell me how well it works for you?

It should work perfect because
If you make a new empty table
and in the keydown sub
you add a new line
scoretext.text = KeyCode
you will see the keyboard value of every key that vpx is either accepting or sending to vpinmame.

Which, technically, allows mapping of staged flips
just no pretty label of course.

Now if @Toxie could UPDATE VPX to SAVE the keyboard assignments to CustomKeys
Because it does not save them, you have to jam them in by hand.
This would be many times easier for people
Then it would just be assigning a key, pick a button, let vpmkeys.vbs know i got actual staged flips and this is the keycode

**Comment by Primetime5k** [comment]
How did you test @wiesshund? Hitting the "U" & "I" key on your keyboard. or a button on a controller/cab?  Still run into the same issue, the 2nd stage input still isn't recognized with this approach, or I'm doing something wrong.

**Comment by wiesshund** [comment]
i manually put the keys into VPinballX.ini (10.8, 10.7 uses registry)
i opened vpx and verified that they keyboard entries showed for the custom keys
i opened windows game controlled config to check what buttons windows called the buttons i was pressing
in vpx, i picked those buttons in the custom keys and closed prefs

I created new blank table
I added a line below line 43 in the script

`ScoreText.Text=KeyCode`

i launched new table

i pressed the controller buttons

the textbox (may have to run in desktop mode to see it)
reported 22 and 23 as i pressed the buttons

i opened up vpmkeys.vbs
and since i dont have the buttons and so have no clue what table i have to pick
i went and assigned the custom keys to coinslot and start game
and fired up a random table

pushed those buttons, which are not bound in VPX to coin or start at all, and BING coins going in and game starting

went in VPMKeys again, assigned them to other things and poof, they work, though not productive but.

And since you asked, i went poking for a table that i have that has stagged flips

I have Twilight Zone

so in VPM Keys

`
' Double leaf/staged flipper switches: By default these are just set to LeftFlipperKey and RightFlipperKey
Dim keyStagedFlipperL : keyStagedFlipperL = 22
Dim keyStagedFlipperR : keyStagedFlipperR = 23`

Now i got staged flips going off my gamepad thumbstick buttons (9 and 10)
also going off if i press U or I

Now, i did NOT go and do this to the table script as i have seen some posts suggesting to do

`Const cSingleLFlip = 0
Const cSingleRFlip = 0

keyStagedFlipperL=""
keyStagedFlipperR=""`

maybe that is outdate or maybe backwards, not sure
but if i do that as a test, staged flips cease working

Now...
Grand Lizard....

The table itself Does NOT support staged flips
I am looking over the wiring schematic for it
It only has one set of flipper contacts per side.

So try twilightzone?

![image](https://github.com/vpinball/vpinball/assets/8684777/b2edd227-2412-4fcd-b089-2d795c6ba7b6)

**Comment by wiesshund** [comment]
If you WANT staged flips on Grandlizard, you could, but you would have to edit the tablescipt and add them custom
the ROM doesnt offer the option

**Comment by Primetime5k** [comment]
Yeah - got dig a little deeper.  Those aren't accurate, as they don't have both upper switches (Thanks for the pic, W!).  FYI - I've basically looked @ every schematic from '78 for any table that has more that three flippers and have modded every single table with them. (Think my popper staged flipper playlist is like 150 tables). I've mastered that part, just never had any luck with the key mapping piece so needed to use J2K.  Will give it another go today.  Thanks again for your effort and help with this!

![Screenshot_20230803_182024_Firefox_Beta](https://github.com/vpinball/vpinball/assets/127700397/5d2f451e-4987-4c2e-aba3-2c37c4ee4a77)

**Comment by wiesshund** [comment]
Grand Lizard wont work 
For one, S11 machine translation has no definition for other than 2 flipper switches
So nothing will happen

And the table itself has no solenoid callback for sURFlipper or sULFlipper
So nothing will happen even if S11 defines it

and running the ROM independently and poking every possible switch I didnt show the callback happening.
does not mean that it is not there, but i could not find it, maybe only works in combo?

For grand lizard to work
S11 needs
Const swURFlip       = switch_number
Const swULFlip       = switch_number

Solcallback is trivial, if there is one to be found.
If one shows up then Grandlizard needs

 SolCallback(sURFlipper) = "URFlipper"
 SolCallback(sULFlipper) = "ULFlipper"

and

 Sub URFlipper(enabled) 'Upper Right FLipper
    If enabled Then
Do dah flipper ting Mon
  End If
End Sub


 Sub ULFlipper(enabled) 'Upper Left Flipper
    If enabled Then
 Do dah flipper ting Mon
  End If
End Sub


Do you happen to know or have a guess at the switch numbers you suspect?


Twilight Zone by the way, should work for you now, no joy2key crap, yes?

**Comment by wiesshund** [comment]
Give me a minute
I may have something for you.

I am assuming you got twilight zone working?

**Comment by wiesshund** [comment]
[s11 MODIFIED.zip](https://github.com/vpinball/vpinball/files/12268123/s11.MODIFIED.zip)
[GRANDLIZARD SCRIPT.zip](https://github.com/vpinball/vpinball/files/12268178/GRANDLIZARD.SCRIPT.zip)


Ok grandlizard.vbs obviously paste into table script editor

s11 modified vbs, back up your original
then drop this one in the scripts folder
then run grandlizard

No voodoo going on
I am running strictly on solenoid callbacks and switches.

changes to table script
added upper flipper call backs
move upper flipper RIGHT rotations to those callbacks

remove hard set flipper keydown/up commands
Only RIGHT upper flipper
Upper LEFT flipper best i can tell is combined

changes to S11
added switch assignments for upper flips
SW 45 and 46

change the keyup/downs for flippers to simply match those of other machine translations that definitely address staged flips
compare to wpc.vbs

The solenoids fire and the flippers work

IF for some reason, you want to artificially make but upper flippers staged
remove the Const cSingleLFlip = 1 at top of script

Then go to 
Sub LFlipper(Enabled)

and comment out the upper flipper rotation lines.

This is of course not how the table would normally be, best i can possibly tell.

FYI
this will work for anyone staged keys or not
i tested by setting mine back to non staged keys

**Comment by wiesshund** [comment]
new machine translation got added to VPX :)

Let me know how your staged flips are working for you?

**Comment by Primetime5k** [comment]
Was just coming to ask about this...
![SmartSelect_20230806_143630_Chrome](https://github.com/vpinball/vpinball/assets/127700397/af8d01b4-bde8-4a68-b9cd-eb2bb1da76d8)

**Comment by wiesshund** [comment]
@Primetime5k i gave the info to 3rdaxis so he can update his table.
Id rather not upload it as a mod just for this

**Comment by LynnInDenver** [comment]
I have both stages flippers and magnasaves on my cabinet. Granted, the staged second flippers are set to the default keys in my KL25Z controller, so as long as a table supports it, I get proper staged flippers without messing with it. But a method to assign new keys to that in the UI, as well as documentation and emulation for manufacturer systems that actually had it in some way or another would be nice.

I mean, yes, it seems to be a lot of work for what seems to be less than a dozen tables in emulation, but it's a lot more common in the modern era (there might be one or two multiple flipper tables that DON'T do staged flippers in the Spike1/Spike2 line).

**Comment by wiesshund** [comment]
> I have both stages flippers and magnasaves on my cabinet. Granted, the staged second flippers are set to the default keys in my KL25Z controller, so as long as a table supports it, I get proper staged flippers without messing with it. But a method to assign new keys to that in the UI, as well as documentation and emulation for manufacturer systems that actually had it in some way or another would be nice.
> 
> I mean, yes, it seems to be a lot of work for what seems to be less than a dozen tables in emulation, but it's a lot more common in the modern era (there might be one or two multiple flipper tables that DON'T do staged flippers in the Spike1/Spike2 line).

Hi Lynn
You can do the above (If using 10.7 you would edit the registry rather than ini) to map staged flipper buttons without having the controller do keyboard emulation (or resorting to joy2key)

I have a request open to change VPX to SAVE the keyboard maps (which is why it doesnt automagically work now) when setting custom keys, also asked to add 2 additional keys.
That would allow for 2 staged flips and 4 digital 4way stick positions.

If you have grand lizard, grab that copy of the grand lizard script, and you will get your staged upper right flipper.
Williams did not stage the upper left, no idea why, but all the schematics say no.

Though if you see my notes above, you can force the left flipper to be staged.

![image](https://github.com/vpinball/vpinball/assets/8684777/2d215ba2-60ee-4347-876f-96c78c7d037f)

**Comment by wiesshund** [comment]
machine translation in beta 10.8 release is named 

s11_grand_lizard.vbs

**Comment by wiesshund** [comment]
I am looking at solar flare which uses System 7 vs system 11

having trouble identifying the switch numbers as they would pertain to vpinmame

**Comment by wiesshund** [comment]
So anyone wanna try out solar flare, with VPM controlled staged flips?

**Comment by Kamun** [comment]
I've got new leaf switches on the way, I can try and test with alternate assignments per the instructions left here soon. 

But to echo my previous comments and the original reason for creating this thread, having the upper flippers be independently assignable via the Keys, Nudge, etc GUI would save so much headache vs trying to set it up via these scripts.

**Comment by wiesshund** [comment]
Oh you wont get away from the scripts
You still have to tell vpmkeys, so it knows what the key means.
Now you can add defaults to vpmkeys of course
i cant say u i o p are good defaults LOL

and the machine translation scripts still have/had to be updated
because the function does not work if the machine translation does not include it.

But i do want to see the urflip and ulflip keys added to the UI so they are internally recognized variables
because....

Dim keyStagedFlipperL : keyStagedFlipperL = keyupperLeft
Dim keyStagedFlipperR : keyStagedFlipperR = keyupperRight

Ok that well those exist in vpmkeys already, right?
They work (if you are a desktop keyboard user)

Except, unless you use joy2key, you cant do a damned thing with it.
Well that negates controllers :(

So need a UI map that equates to those, so you can set a keyboard key (or use the defaults)
So VPX SEES you poking that controller button and knows to tell vpinmame

Just like what we are doing with the 4 custom keys at the moment.

So none of the above is wasted, just need a little help from the UI to have 4 custom keys AND upper flipper keys
Because everyone wants to play baby pacman and granny and the gaters, right?
(which is what i normally do with my custom keys, to assign the 4 way stick)

**Comment by Wylted1** [comment]
> > I have both stages flippers and magnasaves on my cabinet. Granted, the staged second flippers are set to the default keys in my KL25Z controller, so as long as a table supports it, I get proper staged flippers without messing with it. But a method to assign new keys to that in the UI, as well as documentation and emulation for manufacturer systems that actually had it in some way or another would be nice.
> > I mean, yes, it seems to be a lot of work for what seems to be less than a dozen tables in emulation, but it's a lot more common in the modern era (there might be one or two multiple flipper tables that DON'T do staged flippers in the Spike1/Spike2 line).
> 
> Hi Lynn You can do the above (If using 10.7 you would edit the registry rather than ini) to map staged flipper buttons without having the controller do keyboard emulation (or resorting to joy2key)
> 
> I have a request open to change VPX to SAVE the keyboard maps (which is why it doesnt automagically work now) when setting custom keys, also asked to add 2 additional keys. That would allow for 2 staged flips and 4 digital 4way stick positions.
> 
> If you have grand lizard, grab that copy of the grand lizard script, and you will get your staged upper right flipper. Williams did not stage the upper left, no idea why, but all the schematics say no.
> 
> Though if you see my notes above, you can force the left flipper to be staged.
> 
> ![image](https://user-images.githubusercontent.com/8684777/258664480-2d215ba2-60ee-4347-876f-96c78c7d037f.png)

The preliminary schematics don't have the left and the final manual doesn't have the schematic, bit there is this note:
![Screenshot_20230803_182024_Firefox Beta](https://github.com/vpinball/vpinball/assets/25475953/e608ede8-4371-4fe9-ac8f-f2ec920a1951)
And you can see the separate wiring to them on a physical machine

**Comment by Wylted1** [comment]
@Primetime5k made an excellent spreadsheet, which might be useful for cross-referencing machines on the same system when the manuals/schematics aren't enough.

**Comment by passionforpins** [comment]
I just want to bow in wiesshund's  general direction. I've been watching this epic saga and amazed at the systematic way you've solved this "feature" request. You should get a lifetime achievement award for all you've contributed through the years.
Bravo

**Comment by wiesshund** [comment]
> I just want to bow in wiesshund's general direction. I've been watching this epic saga and amazed at the systematic way you've solved this "feature" request. You should get a lifetime achievement award for all you've contributed through the years. Bravo

I would not do much bowing
I dont know how to actually put in the vpx UI and put the variables in the inbuilt environment.

In order for VPX, by itself, to support Every possible table in existence, without resorting to 3rd party software etc
we need VPX to natively understand 2 more keys

Ok 3 technically but i mean NOBODY puts a dollar bill changer in their cab plugged into coinslot 3 so we can forgo that one.

And yes, there are all of 2 tables that use a 4 way analog joystick
So probably 90% of cab owners dont give a fart about being able to map that and would just be happy being able to directly use their staged buttons but

Somebody out there WILL want to play those two tables
AND use their staged flips
AND their magnasaves
And not rely on 3rd party hocus pocus to get it to work.

(Please dont use staged flips on ltd haunted hotel, it's a fast way to end your multiball, you need fully independent use, just throwing that in there)

My reason for looking at these 2 machines, Grand Lizard and Solar Fire is
They are deviants, staged flips is not the norm for their controller type.
So their machine translation file in turn does not support properly operating as staged.

So That lets me show a valid need of either a machine file update OR a machine file alternate in the case of conflicts.
Cause those have to be available in vpx official scripts so that authors\editors can call on them and trust that they will handle their side of things right.

It also lets me show the dev team a real world 100% valid case of VPX does not allow enough button mappings to support these machines which are unique as they use lower flip buttons upper flip buttons and magnasave buttons all in the same cabinet.

And hopefully that helps the cause, because now it isnt just some random feature request that has in in VPX way of working with it, now it is an actual case of we are actually missing something and to work around it is to actually take away from something else.

Turn the idea from Feature request to correcting a Flaw

And i can actually throw them a working example of why this should be

Plus i had to bug report the fact the UI wont even save keyboard assignments the existing customs anyways.

**Comment by wiesshund** [comment]
Not to mention
I cannot possibly go through 145 tables to check reason for staged flips not working.

But if some others see what was done here, it might give them some ideas to run with to find the correct solution for the other machines

Here is Williams Solar Fire (assuming you have already working staged flips or have made them work as discussed above)
Updated machine translation for S7. just a small change.
Will NOT harm any other S7 tables at all
https://drive.google.com/file/d/1C1c34UMyx8cFcsnCMgnLPmXFzGlMMF1W/view?usp=drive_link

Table script update for Goldchicco's 1.2a version of the table.
Cause you can put 16 flipper buttons on the can and have the ROM recognize every single one, but a tables gonna do what the script says to do LOL
https://drive.google.com/file/d/13FIGIVx_Rr6a3RbhTC4ZXpCAeWHGAItj/view?usp=drive_link.


Incidentally, since i obviously have gone off rambling (kidding, well not really but hopefully this helps somebody)
I do not thing the System 7 controller has a bloody clue that a flipper is in play.
I have looked and looked at the wiring, and i think the controller simply kills power to all table action solenoids if game is over
or game is tilted or ball is drained.

The logic hasnt a clue if a flipper was pressed or not, because i see no feedback tracers from flipper to logic at all.

This made S7 easy due to some built in things Pinmame dictates.
PinMAME dictates that
'-- Flipper solenoids (all games)
Const sLRFlipper = 46
Const sLLFlipper = 48
Const sURFlipper = 34
Const sULFlipper = 36

This handles most machines, unless something bizarrely deviates (thanks LTD System IV heavily hacked)

PinMAME also says that these switches, though they do not exist in a real SS machine
will be auto magically internally mapped and correstond to the above solenoid operation

swLRFlip       = 82
swLLFlip       = 84
swURFlip       = 83
swULFlip       = 81

What that means is
If there is no odd deviation in the creation of the real pin
VPM will interpret those switch numbers as flipper switches

VPM will then map them internally to activation of solenoids 
Const sLRFlipper = 46
Const sLLFlipper = 48
Const sURFlipper = 34
Const sULFlipper = 36

so those SolCallbacks will activate the appropriate flipper.

This works fine on many machines because the flippers dont run through the logic
the logic does not know the flipper is doing anything
the logic only cuts power in a global sense when they should not be running.

The fun comes in when said flippers do in fact run through the logic, because then you have to find actual switch IDs and actual solenoid IDs.

LTD Haunted Hotel was like that.
To get 6 flippers on 2 independent sets of controls, on a system board that was not originally designed for it
they stole solenoids and table switches that hotel would not be using, and rewrote how they would work, in the logic itself.
And ran 6 flippers through the logic board in totally non standard circuits.

So in that case, actual real switch ID's and actual real solenoid numbers had to be determined.
to get all of them working As the ROM itself operates them.

Could have been done outside the ROM of course but we preserve these machines better if we make them accurate in operation.
Plus we learn more about them.

**Comment by wiesshund** [comment]
@Wylted1 try applying what i did to Grand Lizard to Highspeed
If it does not work, i may have another idea.

**Comment by Primetime5k** [comment]
Ah...a bit lost here.  What are you looking for?  I have staged flippers working on all tables that have them, so I have the scripts altered already. Interesting point on Elvira - I'll have to take a look.  Did quite a bit from the comfort of my bed by remoting to my cab, scripting, then only seeing if the staging worked.

**Comment by wiesshund** [comment]
> Ah...a bit lost here. What are you looking for? I have staged flippers working on all tables that have them, so I have the scripts altered already. Interesting point on Elvira - I'll have to take a look. Did quite a bit from the comfort of my bed by remoting to my cab, scripting, then only seeing if the staging worked.

I am looking for a list of just ONE table per Controller Type, that is staged flipper outfitted.
As vpx wise it only needs corrected one, for the controller

As far as You having made the tables staged flips
The question is How did you do it?
If it isn't the correct way, regardless of working, it is not of use to other authors or table users.
I can make it work lots of different ways, but only one of them would be considered a "correct" way.

I cant go asking for things to be updated in a non standard way, that stuff is a "youre on your own" kind of thing

Spike1 and 2 dont enter the picture as VPM doesnt emulate them
that is just a matter of defining a variable and making a UI existence of it

**Comment by wiesshund** [comment]
guess no updates here?
I cant go through 150 tables, will take too long

**Comment by wiesshund** [comment]
If anyone care, this is not in the latest build.
But as yet, no one at all has offered to help see what machine translations need addressed.
And i do not have time to download and examine 150 tables

**Comment by wiesshund** [comment]
@Primetime5k  @gear323 @Wylted1 

Any of you guys want to help with narrowing down which machine controllers are used in the staged flip tables?
i only need to know the controller type.

I already have WPC, S11, S7, S6 and have tested those.
And we have the keymapping in place inside VPX already, just forgive the cramped UI

![image](https://github.com/vpinball/vpinball/assets/8684777/a328a1c3-ece7-46e6-94af-7d4630ca6831)

**Comment by wiesshund** [comment]
@Kamun ?

**Comment by wiesshund** [comment]
Jurassic Park, while in the list of staged flips tables is NOT staged flips.

I have gone extensively through the schematics and parts list.
It does not have a staged upper right flip

If you want one, you can do it outside of the ROM
using the key down/up's

StagedLeftFlipperKey and StagedRightFlipperKey
which are now supported, but that table itself does not have staged flippers even though it is in the list.

Same with Guns and Roses, and other of same DE controller.
VPM shows only 1 set of flipper solenoid channels (Representing the Initiate, and Hold coils) per table side.
To actually be staged, an additional set of channels would need to be present but they are not.

If you guys want to offer up a switch number and solenoid number (will actually be 2 adjacent solenoids)
I will look at it again, but i have run through 200 switches and have not gotten any unidentified solenoid pair to appear.

Again, this can now be done in the script if one wants, even if the table does not really have staged flips

just declare a const
Const IsStaged = 1/0

then in the sub for the flipper callback put the staged flip in an if then argument
if IsStaged = 0 then
move the flipper
and if

And then do the keyup/down
If keycode = StagedLeftFlipperKey Then
If IsStaged = 1 then
move flipper
End If
End Iff

But i can not yet find ROM level support for it in VPM

**Comment by wiesshund** [comment]
Controller work so far, and some tables that use said controllers (In script form only)

https://drive.google.com/drive/folders/1Ke4ELUazzZOAbMFVEqYWugRnlW7dhsTl

**Comment by Wylted1** [comment]
> Jurassic Park, while in the list of staged flips tables is NOT staged flips.
> 
> I have gone extensively through the schematics and parts list. It does not have a staged upper right flip
> 
> If you want one, you can do it outside of the ROM using the key down/up's
> 
> StagedLeftFlipperKey and StagedRightFlipperKey which are now supported, but that table itself does not have staged flippers even though it is in the list.
> 
> Same with Guns and Roses, and other of same DE controller. VPM shows only 1 set of flipper solenoid channels (Representing the Initiate, and Hold coils) per table side. To actually be staged, an additional set of channels would need to be present but they are not.
> 
> If you guys want to offer up a switch number and solenoid number (will actually be 2 adjacent solenoids) I will look at it again, but i have run through 200 switches and have not gotten any unidentified solenoid pair to appear.
> 
> Again, this can now be done in the script if one wants, even if the table does not really have staged flips
> 
> just declare a const Const IsStaged = 1/0
> 
> then in the sub for the flipper callback put the staged flip in an if then argument if IsStaged = 0 then move the flipper and if
> 
> And then do the keyup/down If keycode = StagedLeftFlipperKey Then If IsStaged = 1 then move flipper End If End Iff
> 
> But i can not yet find ROM level support for it in VPM

I'm not a programmer so I don't pretend to understand VPM or even all of the code I use, but I do own Tommy, which has staged flippers and uses a separate SS Flipper Board.  Here's what I did, based on @Primetime5k's code:

`Const StagedFlipper = 0			'0 = No staged flipper, 1 = Staged Flipper (Seperate input for Upper Left flipper)

Sub Tommy_KeyDown(ByVal Keycode)
	If keycode = 3 Then Controller.Switch(8)=1                    		'extra ball button
	If keycode = LeftTiltKey Then Nudge 90,1.5 : SoundNudgeLeft
	If keycode = RightTiltKey Then Nudge 270,1.5 : SoundNudgeRight
	If keycode = CenterTiltKey Then Nudge 0,1.5 : SoundNudgeCenter
	If keycode = PlungerKey Then Plunger.Pullback: SoundPlungerPull
	If keycode = LeftFlipperKey Then
		FlipperActivate LeftFlipper, LFPress
		LeftPress = 1
		If StagedFlipper <> 1 Then
			FlipperActivate LeftFlipper1, LFPress1
			SolULFlipper True
		End If
	End If
	If StagedFlipper = 1 Then
		If keycode = KeyUpperLeft Then
			FlipperActivate LeftFlipper, LFPress
			FlipperActivate LeftFlipper1, LFPress1
			SolLFlipper True
			SolULFlipper True
		End If
	End If
	If keycode = RightFlipperKey Then FlipperActivate RightFlipper, RFPress
	If keycode = StartGameKey Then SoundStartButton
	If keycode = AddCreditKey or keycode = AddCreditKey2 Then
		Select Case Int(rnd*3)
			Case 0: PlaySound ("Coin_In_1"), 0, CoinSoundLevel, 0, 0.25
			Case 1: PlaySound ("Coin_In_2"), 0, CoinSoundLevel, 0, 0.25
			Case 2: PlaySound ("Coin_In_3"), 0, CoinSoundLevel, 0, 0.25
		End Select
	End If
	If VRRoom > 0 and VRRoom < 3 Then
		If keycode = LeftFlipperKey Then
			VRFlipperButtonLeft.x = VRFlipperButtonLeft.x + 5
		End If
		If keycode = RightFlipperKey Then
			VRFlipperButtonRight.x = VRFlipperButtonRight.x - 5
		End If
		If keycode = StartGameKey Then
			LaunchButton.y = LaunchButton.y - 2
			Flasher2.y = Flasher2.y - 2
			Flasher3.y = Flasher3.y - 2
		End If
		If keycode = 3 Then
			ExtraBallButton.y = ExtraBallButton.y - 2
			Flasher1.y = Flasher1.y - 2
		End If
		If keycode = PlungerKey Then
			TimerVRPlunger.Enabled = True
			TimerVRPlunger2.Enabled = False
		End If
	End If
	If vpmKeyDown(keycode) Then Exit Sub
End Sub

Sub Tommy_KeyUp(ByVal Keycode)
	If keycode = 3 Then Controller.Switch(8)=0
	If KeyCode = PlungerKey Then
		Plunger.Fire
		If BIPL Then
			SoundPlungerReleaseBall()			'Plunger release sound when there is a ball in shooter lane
		Else
			SoundPlungerReleaseNoBall()			'Plunger release sound when there is no ball in shooter lane
		End If
	End If
	If keycode = LeftFlipperKey Then
		LeftPress = 0
		FlipperDeActivate LeftFlipper, LFPress
		FlipperDeActivate LeftFlipper1, LFPress1
		SolULFlipper False		
	End If
	If StagedFlipper = 1 Then
		If keycode = KeyUpperLeft Then
			FlipperDeActivate LeftFlipper1, LFPress1
			SolULFlipper False
			If LeftPress = 0 Then
				FlipperDeActivate LeftFlipper, LFPress
				SolLFlipper False
			End If
		End If
	End If
	If keycode = RightFlipperKey Then FlipperDeActivate RightFlipper, RFPress
	If VRRoom > 0 and VRRoom < 3 Then
		If keycode = LeftFlipperKey Or keycode = KeyUpperLeft Then
			VRFlipperButtonLeft.x = VRFlipperButtonLeft.x - 5
		End If
		If keycode = RightFlipperKey Then
			VRFlipperButtonRight.x = VRFlipperButtonRight.x + 5
		End If
		If keycode = StartGameKey Then
			LaunchButton.y = LaunchButton.y + 2
			Flasher2.y = Flasher2.y + 2
			Flasher3.y = Flasher3.y + 2
		End If
		If keycode = 3 Then
			ExtraBallButton.y = ExtraBallButton.y + 2
			Flasher1.y = Flasher1.y + 2
		End If
		If keycode = PlungerKey Then
			TimerVRPlunger.Enabled = False
			TimerVRPlunger2.Enabled = True
		End If
	End If
	If vpmKeyUp(keycode) Then Exit Sub
End Sub
Dim bFlippersEnabled : bFlippersEnabled=False

Sub SolEnableFlips(Enabled)
	bFlippersEnabled = Enabled
End Sub

Sub SolLFlipper(Enabled)
	If Enabled And bFlippersEnabled Then
		LF.Fire
		If leftflipper.currentangle < leftflipper.endangle + ReflipAngle Then 
			RandomSoundReflipUpLeft LeftFlipper
		Else 
			SoundFlipperUpAttackLeft LeftFlipper
			RandomSoundFlipperUpLeft LeftFlipper
		End If		
	Else
		LeftFlipper.RotateToStart
		If LeftFlipper.currentangle < LeftFlipper.startAngle - 5 Then
			RandomSoundFlipperDownLeft LeftFlipper
		End If
		FlipperLeftHitParm = FlipperUpSoundLevel
	End If
End Sub

Sub SolRFlipper(Enabled)
	If Enabled And bFlippersEnabled Then
		RF.Fire
		If rightflipper.currentangle > rightflipper.endangle - ReflipAngle Then
			RandomSoundReflipUpRight RightFlipper
		Else 
			SoundFlipperUpAttackRight RightFlipper
			RandomSoundFlipperUpRight RightFlipper
		End If
	Else
		RightFlipper.RotateToStart
		If RightFlipper.currentangle > RightFlipper.startAngle + 5 Then
			RandomSoundFlipperDownRight RightFlipper
		End If	
		FlipperRightHitParm = FlipperUpSoundLevel
	End If
End Sub

Sub SolULFlipper(Enabled)
	If Enabled And bFlippersEnabled Then
		LeftFlipper1.RotateToEnd
	Else
		LeftFlipper1.RotateToStart
		FlipperLeft1HitParm = FlipperUpSoundLevel
	End If
End Sub`

I don't know GnR as well, but it appears to use the same flipper power board and has code that is probably better than mine:

`Sub SolULFlipper(Enabled)
	If Enabled Then
		LeftFlipper1.RotateToEnd
		If leftflipper1.currentangle < leftflipper1.endangle + ReflipAngle Then 
			RandomSoundReflipUpLeft LeftFlipper1
		Else 
			SoundFlipperUpAttackLeft LeftFlipper1
			RandomSoundFlipperUpLeft LeftFlipper1
		End If		
	Else
		LeftFlipper1.RotateToStart
		If LeftFlipper1.currentangle < LeftFlipper1.startAngle - 5 Then
			RandomSoundFlipperDownLeft LeftFlipper1
		End If
		FlipperLeft1HitParm = FlipperUpSoundLevel
	End If
End Sub`

**Comment by Wylted1** [comment]
![Screenshot_20230828_123808_Drive](https://github.com/vpinball/vpinball/assets/25475953/d9d43e66-52d1-462c-9355-3510fb47b188)
![Screenshot_20230828_125813_Drive](https://github.com/vpinball/vpinball/assets/25475953/6c1f0378-6d37-4a2f-b2fd-f42b890c9216)

**Comment by wiesshund** [comment]
that is only showing flipper solenoids
Most multi flipper tables use individuals per flipper.

Look at tommy's switch matrix and wiring schematic though
![image](https://github.com/vpinball/vpinball/assets/8684777/e4a86e19-f309-4e62-bf01-fc1d98b96617)

And this part number, appears to be a single leaf switch
![image](https://github.com/vpinball/vpinball/assets/8684777/5b85a942-f4a2-4cb3-a3df-36509393fb6a)
Though the manual later contradicts itself on that.

look at manual pages 61 62 and 63.
Only a single channel per side shown.

If it is controlled by some kind of external cutout circuit, it is not represented in the wiring diagram
and VPM basically won't know a thing about it.
The only way it would is to open a VPM ticket, because you need some assumed switches (those can be anything outside the matrix) and those need to be tied to solenoids

Const sURFlipper = 34 
Const sULFlipper = 36 

These are VPM defaults ^^^^

But those do not work in All cases, because some tables are deviations, where they literally steal something that is a standard solenoid for something else, default VPM callback for main flippers is 46 and 48.
In the case of this machine, 46 and 47 fire together (This is normally activate and hold) on a single keypress
45 and 46 also fire together, that is the 2 windings in the main flipper solenoids, move and hold.

If i close the single left flipper switch, solenoid channels engage
Main left Move
Main left Hold (EOS)

Same with right.

You would need VPM team to examine and see there, because there isnt anything ROM wise to hinge the staged flip from.
You can mimic an external control, like below pretty simply

The VPX Table is coded to trigger the upper left flipper off the engage winding of the main left flipper
That is not really correct.

 you could do this (Or any variation that suits you)
 Table will then work for anyone, staged flips or not, just by having their VPM keys correct
 Having staged flips mapped in key prefs (key AND button)
 and setting the const to reflect what they have.
 
 NOTE: it is not the best idea to steal the solenoid call for the main flipper to hold the staged flipper.
 This table it seems ok, but some machines turn off the solenoids, which could result in either a laggy staged flipper
 or a staged flipper that keeps falling down, this ROM seems ok, but i know some that are not.

'Cab Const HasStaged  (0 = no 1 = yes)

Const HasStaged = 1
Dim FActive

'table key Downs
If Keycode = StagedLeftFlipperKey Then
	If HasStaged = 1 Then
		If FActive = 1 then
			LeftFlipper1.RotateToEnd:PlaySound SoundFX("fx_flipper2",DOFContactors)
		End If
	End if
End If

'table key Ups

If Keycode = StagedLeftFlipperKey Then
	If HasStaged = 1 Then
LeftFlipper1.RotateToStart:PlaySound SoundFX("fx_flipperdown",DOFContactors)
	End If
End If

Sub SolLFlipper(Enabled)
If Enabled Then
FActive = 1
LeftFlipper.RotateToEnd:PlaySound SoundFX("fx_flipper1",DOFContactors)
Else
FActive = 0
LeftFlipper.RotateToStart:PlaySound SoundFX("fx_flipperdown",DOFContactors)
End If
End Sub

Sub SolULFlipper(Enabled)
	If HasStaged = 0 then
		If Enabled Then
		LeftFlipper1.RotateToEnd:PlaySound SoundFX("fx_flipper2",DOFContactors)
		Else
		LeftFlipper1.RotateToStart:PlaySound SoundFX("fx_flipperdown",DOFContactors)
		end if
	End If
End Sub

**Comment by Wylted1** [comment]
@wiesshund
Apparently DE.vbs already has vpm keys for staged flippers in JP and some others, but not all 5?

Also gathered up the solid state flipper board info and schematics for all DE machines with staged flippers (mostly duplicated game-to-game):
JP p66-67
TftC p82-85
Tommy p81-84
WWF p81a-85
GnR p87-89

And a couple quick pictures showing the 3 flipper switches in JP and GnR
![Screenshot_20230829_033431_Drive](https://github.com/vpinball/vpinball/assets/25475953/98529b23-80ea-4d51-bbeb-4981784ec772)
![Screenshot_20230829_033614_Drive](https://github.com/vpinball/vpinball/assets/25475953/4273b785-a94c-4f91-8002-95223718bcca)

**Comment by Wylted1** [comment]
Looked at the earlier DE flipper schematics as well:
Secret Service p24-27
Torpedo Alley p26-27, 37-38
And Maverick, because why not:  p70, 90-92

**Comment by wiesshund** [comment]
@Wylted1 
It isnt in the logic switch matrix (which means VPM wont know about it)

So where does the control go to?
If it goes to some external splitter board, then like the external sound boards, it probably wont be represented in VPM.

If that is the case, you would want to show the VPM team, and see if that is something that could be built into the DataEast emulation, or if doing it via script is the only way.

You said you had a real Tommy?
Would it be possible for you to carefully trace out how it is set up?

Does it use an add-on board to facilitate the 3rd flipper?
I mean it is entirely possible to run one flipper into a secondary board, let normal flipper control pass through to the main flipper, and tap the voltage, then use a simple control to flip the 3rd flipper of a 3rd button, entirely outside of the logic.

If it does something like that, i would think it would be possible to implement into VPM
as you are simply saying that 
If flippers are powered  then Switch ## will result in activation of Solenoid##
and then you use a switch number and solenoid number outside of the logic matrix.

That is if you want it to work within VPM, which is simpler from an author standpoint because it is scriptless
aside from simply saying if solcallback then flip, if not unflip
Not to mention it is the difference between "i can make this work" and "this works right."

**Comment by wiesshund** [comment]
@Wylted1 
Im looking at Tommy parts, sadly i cant find pictures for context but
there appears to be a solid state transistor controlled independent flipper board ?

it looks like only 1 used? but like i said i cant find pictures for context.

If that is so, it could be where it derives it's staged flipper control from?
Which may in fact place it outside of the ROM's knowledge.

If i understand right you have the Tommy table, maybe you could confirm that, without harming your machine?

If that is the case, we can ask vpm team if it is possible to mimic that?
I think it should be possible.

**Comment by Wylted1** [comment]
@wiesshund here's the manual, a picture of the board, and a picture of the schematic for it (p.82).  The SS flippers info starts on page...80.5 (p.105 of the pdf).  The first page references the Upper Right flipper and a 9v hold instead of UL and 8v because it was just copied from TftC.

https://www.ipdb.org/files/2579/Data_East_1994_The_Who_s_Tommy_Pinball_Wizard_Manual.pdf
![20230830_021115](https://github.com/vpinball/vpinball/assets/25475953/0082a428-eb59-41e5-a128-71a94c623a05)
![20230829_230322](https://github.com/vpinball/vpinball/assets/25475953/ba93f019-ce77-457c-9773-93ead5c76711)

**Comment by wiesshund** [comment]
funky, but best i can tell, the logic hasnt a clue what is going on there?

meaning since VPM does not seem to emulate this, a VPM request or bug needs filed.
can you open one to look at dataeast emulation?

**Comment by Wylted1** [comment]
> funky, but best i can tell, the logic hasnt a clue what is going on there?
> 
> meaning since VPM does not seem to emulate this, a VPM request or bug needs filed. can you open one to look at dataeast emulation?

I can confirm that the ROM does not see the UL flipper switch in the switch test.  From what I can tell, the SS flipper board contains all the logic for the flippers, sending the two primary flipper signals back to the ROM, and switching to the hold voltage (see: the EoS theory of operation).  It basically "unlocks" the electrical path for the flippers ahead of time with Sol #23 (which is emulated in VPM) as a dedicated flipper enable relay, so there is no waiting on communication with the ROM when the flipper switch is engaged (as small as that is); even for the upper flipper there is only *technically* a small delay as the button travels through the gap between the staged leafs.  I could be wrong, though.

**Comment by wiesshund** [comment]
Sol 23 is basically the GameOn solenoid then, for all practical purposes, at least sounds like?
It will only be on when gameplay is valid.
Might be useful for fastflips?

Ok, so what we need to do here, if you want accurate representation, is we need to present this to the VPM team
and see if VPM can emulate the flipper board, which need not really be complex as we need it to emulate a switch and solenoid combo for the staged flips (I would emulate for both left and right, just to cover both possibilities)

I know i can do it very simply via table script now, with a simple check to ensure the main flip was activated, but it is not accurate, and we do kind of want things to be as accurate as we can possibly make them.

This should not be difficult, i dont think? as there are other machine in which the flippers never go through the logic at all
and all the logic does is kill a master power relay, killing power to flippers bumpers and slingshots when game over or tilted.
Logic itself actually has no clue if or when the flippers are being used.
So i would think the means of doing this already exist and simply need applied to the DE emulation, with a dedicated solenoid number assigned to it, one that is out of the matrix of the normal ROM controlled solenoids, then it just needs to know that if upper flipper keypress is sent, as long as main flipper is engaged, activate Solenoid 100 or whatever number.

**Comment by wiesshund** [comment]
@Wylted1 im discussing the Data East 1 controlled machines with vpm team to see what can be done.

Something i want to note, for anyone following this.

Some machines DO use multiple flipper contacts
But they are multiple High Voltage contacts.
The intention is not that those be staged, but that you can only draw so many amps across a single contact.
Those machines should not function as staged, doing so is an indication of a worn or poorly aligned switch stack.
And for VPM's part, VPM will most likely only recognize such a machine(ROM) as a single set of switches closing, and a single set of solenoids operating.

(It's in some of the maintenance manuals, the importance of properly adjusting them so the close evenly together)

F4 Tomcat looks to be such a machine, though i am not the pinball schematics expert.
It has multi contacts on both left and right flipper circuits, it also appears to be high voltage.
If true, VPM most likely would not recognize such a thing as staged, but one could definitely script it as staged even if inaccurate.

This has no bearing on the DE thing, those are staged, the high voltage is handled by that controller board.
I am just mentioning it in case someone sees two switches on a given machine and cant figure out why absolutely nothing works.

**Comment by passionforpins** [comment]
as posted on VPF, that's correct wiesshund. Thanks again for the amazing work you've done bring it into the virtual pinball world.

**Comment by wiesshund** [comment]
> as posted on VPF, that's correct wiesshund. Thanks again for the amazing work you've done bring it into the virtual pinball world.

I will assume this is Clyde and this is regarding use of dual contacts in high voltage flipper circuits?

**Comment by passionforpins** [comment]
Cliffy

**Comment by wiesshund** [comment]
> Cliffy

What's a couple letters among friends
LOL

**Comment by wiesshund** [comment]
Seems participation here has kind of died, but if anyone is still watching, check
https://github.com/vpinball/vpinball/issues/666

I just submitted my work on the issue, as far as the machine translation goes.
The VPX key mapping aspect and scripting was already done a number of revisions ago.

**Comment by toxieainc** [comment]
@Kamun Do you think this can be closed then? Thanks @wiesshund !

**Comment by Kamun** [comment]
I would prefer the option be made available still to bind the upper left / upper right flippers via the GUI, as this was the original reason for the opening of this ticket. I haven't kept up on the schematics side of things as I do not know the deeper intricacies of these systems for the most part but if the keys are going to be directly bindable via GUI then this can be closed, otherwise I would prefer the issue still be raised :)

**Comment by wiesshund** [comment]
> I would prefer the option be made available still to bind the upper left / upper right flippers via the GUI, as this was the original reason for the opening of this ticket. I haven't kept up on the schematics side of things as I do not know the deeper intricacies of these systems for the most part but if the keys are going to be directly bindable via GUI then this can be closed, otherwise I would prefer the issue still be raised :)

It already is, the keys are already in the UI
pull latest build of 10.8

**Comment by Kamun** [comment]
Then we should be good to close :)

**Comment by bbecausereasonss** [comment]
Has anyone had success setting the analog RT/LT on an Xbox One controller to control staged-flipper function? I'm curious, ie) weak flip strong flip? Maybe a joy emulation util is required to set intervals as separate buttons as a work around?

**Comment by wiesshund** [comment]
> Has anyone had success setting the analog RT/LT on an Xbox One controller to control staged-flipper function? I'm curious, ie) weak flip strong flip? Maybe a joy emulation util is required to set intervals as separate buttons as a work around?

Pinball machines dont have analog flippers
They are digital, on/off.
Only way you get weak flip from the buttons is by quickly tapping and releasing before flipper moves full travel.

You can use the triggers, just use antimicroX and remap the trippers to the staged flipper keys.

If you want to use ONLY them as the staged keys, i dont know any gamepads anymore that allow for axis quarant mapping.
Saitek gamepads did, but those are long gone.
You could like set 25% key_A 50% key_B etc

You may be able to do that with joystickGremlinEX?
Not 100% sure but it looks like it.
That is about all i can suggest if you are trying to replicate the staged buttons

I just use the shoulder buttons AND triggers
the whol trying to float the damned button thing is too fiddly for me, even on a real pin
